crates/bevy_pbr/src/environment_map/generate_from_skybox.rs

```diff
@@ -0,0 +1,691 @@
+// https://research.activision.com/publications/archives/fast-filtering-of-reflection-probes
+
+use super::{
+    filter_coefficents::{FilterCoefficentsType, FILTER_COEFFICENTS},
+    EnvironmentMapLight, DIFFUSE_CONVOLUTION_SHADER_HANDLE, DOWNSAMPLE_SHADER_HANDLE,
+    FILTER_SHADER_HANDLE,
+};
+use bevy_asset::{Assets, Handle};
+use bevy_core_pipeline::Skybox;
+use bevy_ecs::{
+    prelude::{Component, Entity},
+    query::{QueryItem, Without},
+    system::{Commands, Query, Res, ResMut, Resource},
+    world::{FromWorld, World},
+};
+use bevy_reflect::Reflect;
+use bevy_render::{
+    extract_component::ExtractComponent,
+    render_asset::RenderAssets,
+    render_graph::{NodeRunError, RenderGraphContext, ViewNode},
+    render_resource::{
+        BindGroup, BindGroupDescriptor, BindGroupEntry, BindGroupLayout, BindGroupLayoutDescriptor,
+        BindGroupLayoutEntry, BindingResource, BindingType, Buffer, BufferBindingType,
+        CachedComputePipelineId, ComputePassDescriptor, ComputePipelineDescriptor, Extent3d,
+        FilterMode, PipelineCache, SamplerBindingType, SamplerDescriptor, ShaderStages, ShaderType,
+        StorageTextureAccess, TextureAspect, TextureDescriptor, TextureDimension, TextureFormat,
+        TextureSampleType, TextureUsages, TextureView, TextureViewDescriptor, TextureViewDimension,
+        UniformBuffer,
+    },
+    renderer::{RenderContext, RenderDevice, RenderQueue},
+    texture::{GpuImage, Image, ImageSampler, Volume},
+};
+use bevy_utils::default;
+
+/// Automatically generate an [`EnvironmentMapLight`] from a [`Skybox`].
+///
+/// Usage:
+/// * Add this component via `GenerateEnvironmentMapLight::default()` to an entity with a [`Skybox`] component.
+/// * The first frame this component is added to the skybox entity, an [`EnvironmentMapLight`]
+/// component will be generated and added to the skybox entity.
```

Comment:
It would be good to give a general idea of how much GPU time this takes. Even something like "takes approximately N ms to filter a 1024x1024x6 cubemap on YYY GPU" would give a hint about whether users need to worry about a frame stutter while the filtering is happening

Reply:
0.46ms GPU time for a 256x256x6 skybox on an RTX 3080. Significant speedups would be gained by copying the first specular mip instead of filtering it, and possibly by using the newer GGX importance sampling technique. I'll leave those improvements for the future/other people, though.

Reply:
Might be worth adding that stat to the doc comment!

Reply:
I'm leery on specific stats, because I don't know how it will compare across GPUs. Different vendors, igpu vs dgpu, different architectures, etc. I don't know how representative this is overall.